HIP/hip-376.md

```diff
@@ -0,0 +1,155 @@
+title: Support for Approve/Allowance and ERC20/ERC721 transferFrom in HTS precompile
+author: Stoyan Panayotov <stoyan.panayotov@limechain.tech>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Review
+created: 2022-03-02
+discussions-to: 
+updated: 
+requires: 206, 218
+---
+
+## Abstract
+
+Extend the functionality from HIP-218 for ERC20 and ERC721 funtions working on
+HTS tokens with support for approve/allowance and transferFrom functions.
+
+## Motivation
+
+The goal of this HIP is to enable users of the HSCS to interact with Hedera
+native tokens in a manner that is similar to how they would in other EVM based
+systems. 
+
+## Rationale
+
+The standard model of ERC-20 and ERC-721 interactions involves directly calling
+the account address. Solidity, however, does a code size check before calling a
+smart contract and reverts if it finds no code. To address this a simple proxy
+redirect is proposed to be added to each token contract accessed that will
+redirect the call (including information about what address was called). HSCS
+will report all token address have this precompile, permitting solidity to
+interact with the token addresses as though it was an account with code.
+
+## User stories
+
+As a hedera smart contract services user, I want to be able to access my HTS
+fungible tokens like an Ethereum ERC-20 token.
+
+As a hedera smart contract services user, I want to be able to access my HTS
+non-fungible tokens like an Ethereum ERC-721 token.
+
+As a hedera smart contract services user, I want to be able to access metadata
+provided by the getTokenInfo in a smart contract.
+
+## Specification
+
+### EVM Precompile
+
+The HTS precompile described in HIP-206, at address `0x167`,will be extended to
+recognize one more function call: `redirectForToken(address,bytes)`. 
+
+### ERC-20 calls directly to Token Accounts
+
+To enable HTS tokens to be more broadly used in DeFi applications the
+interactions with HSCS and Token accounts will be enhanced to allow HTS accounts
+to receive smart contract calls that will be immediately delegated to the HTS
+precompile for evaluation.
+
+#### Redirect contract
+
+Each HTS account will expose a standard redirection EVM bytestream. This
+bytestream will be available to `EXTCODEHASH`, `EXTCODESIZE` and `EXTCODECOPY`
+operations as though the smart contract were deployed itself. The EVM will
+execute the bytecode, which will wrap the call data with the address of the HTS
+Token, to a call to teh HTS precompile. The HTS precompile will then either
+execute or reject the call based on `IERC20` mappings described in the next
+section.
+
+(The following pseudocode will be replaced with real EVM opcodes prior to
+review)
+> ```
+>
+>// Put CALLER in memory
+>// Put the remaining CALLDATA into memory
+>// DELEGATECALL 0x127 with all remaining gas
+>// if the DELEGATECALL failed REVERT()
+>// RETURN the exact same results as the DELEGATECALL
+>
+>```
+
+#### Gas Schedule
```

Comment:
Now that we know what we did for HIP-206 and HIP-218 gas-wise I think we should add a table for each "function group" and the gas price we expect, including the assumptions used to calculate them.

We take the canonical max calls per second, and divide that into the 15MM gas per second, and add a 20% premium.  The trick is figuring out which max TPX measure to use.

We also want to think of the event emissions cost. My thought is we include those in the precompile gas cost.